src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -826,8 +824,44 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_WriteThreadNetw
     break;
 
     case ThreadNetworkDiagnostics::Attributes::RoutingRole::Id: {
-        otDeviceRole role = otThreadGetDeviceRole(mOTInst);
-        err               = encoder.Encode(static_cast<uint8_t>(role));
+        ThreadNetworkDiagnostics::RoutingRole routingRole;
+        otDeviceRole otRole = otThreadGetDeviceRole(mOTInst);
+
+        if (otRole == OT_DEVICE_ROLE_DISABLED)
+        {
+            routingRole = EMBER_ZCL_ROUTING_ROLE_UNSPECIFIED;
+        }
+        else if (otRole == OT_DEVICE_ROLE_DETACHED)
+        {
+            routingRole = EMBER_ZCL_ROUTING_ROLE_UNASSIGNED;
+        }
+        else if (otRole == OT_DEVICE_ROLE_ROUTER)
+        {
+            routingRole = EMBER_ZCL_ROUTING_ROLE_ROUTER;
+        }
+        else if (otRole == OT_DEVICE_ROLE_LEADER)
+        {
+            routingRole = EMBER_ZCL_ROUTING_ROLE_LEADER;
+        }
+        else if (otRole == OT_DEVICE_ROLE_CHILD)
+        {
+            otLinkModeConfig linkMode = otThreadGetLinkMode(mOTInst);
+
+            if (linkMode.mNetworkData)
```

Comment:
This is not correct since Network Data bit only indicates that device requires the Full Network Data.

You can use `otThreadIsRouterEligible` API to check if the device is Router Eligible (REED) and if not, then use `mRxOnWhenIdle` bit.

Reply:
But only the REED requires Full network Data. Else you are already a Router or a Leader.
So if your roule is Child and mNetworkData=1 you are a REED

```
Full Thread Device (FTD) Router
otDeviceRole=OT_DEVICE_ROLE_ROUTER
mRxOnWhenIdle=1
mDeviceType=1
mNetworkData=1

Full Thread Device (FTD) REED
otDeviceRole=OT_DEVICE_ROLE_CHILD
mRxOnWhenIdle=1
mDeviceType=1
mNetworkData=1

Full Thread End Device (FED)
otDeviceRole=OT_DEVICE_ROLE_CHILD
mRxOnWhenIdle=1
mDeviceType=1
mNetworkData=0

Minimal Thread End Device:
otDeviceRole=OT_DEVICE_ROLE_CHILD
mRxOnWhenIdle=1
mDeviceType=0
mNetworkData=0

Sleepy End Device:
otDeviceRole=OT_DEVICE_ROLE_CHILD
mRxOnWhenIdle=0
mDeviceType=0
mNetworkData=0
```

Reply:
It is true that REED must set this flag to '1', but regular FED, MED or even SED can also set it to 1 in practice, so this bit does not indicate if you are a REED or not.

Reply:
`otThreadIsRouterEligible` isn't available on MTD devices. It is causing a build failure on those.

Reply:
Ok re-arranged the logic to take that into consideration.
So The device can only be a REED if `linkMode.mRxOnWhenIdle` is true, `CHIP_DEVICE_CONFIG_THREAD_FTD` is set  and the device is `otThreadIsRouterEligible`

This is correct from both the issue you raised and MTD not having that API. MTD can't be REED anyway.